src/utils/LimitTimer.hxx

```diff
@@ -0,0 +1,132 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file LimitTimer.hxx
+ *
+ * Limits the number of updates per unit time. Initial version written by
+ * Balazs Racz, modified for generalization by Stuart Baker
+ *
+ * @author Balazs Racz, Stuart Baker
+ * @date 9 January 2021
+ */
+
+#include "executor/Timer.hxx"
+ 
+/// This timer takes care of limiting the number of speed updates we send
+/// out in a second. It is a token bucket filter.
+class LimitTimer : public Timer
+{
+public:
+    /// Constructor.
+    /// @param ex executor to run on
+    /// @param update_delay_msec cooldown time delay in milliseconds
+    /// @param max_tokens number of available tokens
+    /// @param callback callback called once after cooldown time delay
+    LimitTimer(ExecutorBase *ex, uint16_t update_delay_msec, uint8_t max_tokens,
+               std::function<void()> callback)
+        : Timer(ex->active_timers())
+        , updateDelayMsec_(update_delay_msec)
+        , bucket_(max_tokens)
+        , bucketMax_(max_tokens)
+        , callback_(callback)
+    {
+        HASSERT(callback);
+    }
+
+    /// Destructor.
+    ~LimitTimer()
+    {
+        cancel();
+    }
+
+    /// Attempts to take a token out of the bucket.
+    /// @return true if the take is successful, false if there are no available
+    ///         tokens, in which case there will be an internal key event
+    ///         generated when tokens become available.
+    bool try_take()
+    {
+        if (bucket_ == bucketMax_)
+        {
+            start(MSEC_TO_NSEC(updateDelayMsec_));
+        }
+        if (bucket_ > 0)
+        {
+            --bucket_;
+            return true;
+        }
+        else
+        {
+            needUpdate_ = 1;
+            return false;
+        }
+    }
+
+    /// Takes one entry from the bucket, and does not give a callback if
+    /// there is no entry left.
+    void take_no_callback()
+    {
+        if (bucket_ > 0)
+        {
+            --bucket_;
+        }
+    }
+
+private:
+    /// Callback from the timer infrastructure. Called periodically.
+    long long timeout() override
+    {
+        ++bucket_;
+        if (needUpdate_)
+        {
+            callback_();
+            needUpdate_ = 0;
+        }
+        if (bucket_ >= bucketMax_)
+        {
+            return NONE;
+        }
+        else
+        {
+            return RESTART;
+        }
+    }
+
+    /// cooldown delay in msec
+    uint16_t updateDelayMsec_;
+
+    /// number of available tokens
+    uint8_t bucket_ ;
```

Comment:
could you squeeze these into an `uint16_t bucket_ : 7;` , make bucketMax the same, and needUpdate with just one bit?
Then the data fields fit into a single 32 bit memory. I think we may have up to 62 of these in RAM.

Reply:
You are correct, there are up to 61 of these in one of our use cases (index 0 is broadcast, index 1 is reserved). I had considered this myself, and since you are are board, let's do it. Fixed.

Don't forget that we have an std::function, so we don't quite fit into 32-bits of memory.